### PR TITLE
[BUGFIX] Fix AmpCast for float16

### PR DESCRIPTION
**Pull request by anko-intel**
## Description ##
OneDNN doesn't support float16 format, so fallback to standard
implementation is needed.
It fixes issue #19631.

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by anko-intel** [comment]
@rongzha1 - could you review?

File: src/operator/tensor/amp_cast.cc [hunk 1]
```diff
@@ -41,25 +41,29 @@ static void AMPCastExCPU(const nnvm::NodeAttrs& attrs,
...
-  mkldnn::engine cpu_engine = mxnet::CpuEngine::Get()->get_engine();
   auto data = inputs[0];
-  if (data.IsView() && data.IsMKLDNNData())
-    data = data.Reorder2Default();
-  const auto i_mem = data.GetMKLDNNData();
-  const size_t i_ndim = data.shape().ndim();
-  mkldnn::memory::dims i_dims = mkldnn::memory::dims(i_ndim);
-  for (size_t i = 0; i < i_ndim; i++) {
-    i_dims[i] = static_cast<int>(data.shape()[i]);
+  if (data.dtype() != mshadow::kFloat16) {
```
Comment (rongzha1) [inline src/operator/tensor/amp_cast.cc:45]: shall we add isValidMKLDNNDataType() to check whether it is supported by mkldnn?  mshadow has so many data types and some of them are not supported.  https://github.com/apache/incubator-mxnet/blob/64f737cdd59fe88d2c5b479f25d011c5156b6a8a/3rdparty/mshadow/mshadow/base.h#L364:3

Comment (anko-intel) [reply src/operator/tensor/amp_cast.cc:45]: I considered that. If  created isValidMKLDNNDataType() function could be used in many places like MKLDNNStorageType() for FInferStorageType it makes sense. But in this particular situation, amp_cast operator only accept 3 float types (see https://github.com/apache/incubator-mxnet/blob/v1.x/src/operator/tensor/amp_cast.h#L70 ) so I just excluded float16  as not supported in MKLDNN.

Comment (rongzha1) [reply src/operator/tensor/amp_cast.cc:45]: OK. LGTM

**Review by rongzha1 — APPROVED** [review]

**Comment by anko-intel** [comment]
@PatricZhao, @szha could you review and merge if everything is ok?

**Review by szha — COMMENTED** [review]
Thanks for the fix! Could you add a test for verification?

**Comment by anko-intel** [comment]
Hi @szha,
Originally I thought that float16 was not intended to be passed to the amp_cast in CPU context and I treated this change as fix only for  #19631.
Now I have enabled existing test for float32->float16 for amp_cast on CPU.

**Comment by anko-intel** [comment]
@mxnet-bot run ci [centos-cpu, unix-gpu]

**Pull request merged**